OCPEDGE-1191: feat: initial arbiter cluster enhancement (#1674)
openshift/enhancements: enhancements/arbiter-clusters.md

---
title: arbiter-clusters
authors:
  - "@eggfoobar"
reviewers:
  - "@tjungblu"
  - "@patrickdillon"

>   - "@williamcaban"

Reviewer: william is no longer involved in controlplane, please tag Ramon Acedo

that consists of 2 normal sized nodes, and 1 node that can be less powerful than
the recommended node size. This 1 arbiter node will only be running critical
components for maintaining HA to allow the arbiter node size to be as small and
as low cost as possible with in reason.

## Motivation

Customers at the edge are requiring a more economical solution for HA
deployments at the edge. They can support running 2 node clusters for redundancy
but would like the option to deploy a lower cost node as an arbiter to supply
the 3 nodes for ETCD quorum.

### User Stories

- As a solutions architect for a retail organization, I want to deploy OpenShift
  at n number of store locations at the edge with only 2 regular sized nodes and
  1 lower cost node to maintain HA and keep compute costs down.
- As a solutions architect for cloud infrastructures, I want to offer low cost
  OpenShift deployments on purpose built hardware for a 2 + 1 configuration.
- As an OpenShift cluster admin I want non-critical applications deployed to my
  2 + 1 arbiter node cluster to not be scheduled to run on the arbiter node.

### Goals

> - Provide a new node arbiter role type that supports HA but is not a full master

Reviewer 2: ```suggestion
- Provide a new arbiter node role type that achieves HA but does not act as a full master node
```

> - Provide a new node arbiter role type that supports HA but is not a full master

Reviewer 3: That's a clear goal. We may get some questions around what pieces are HA (e.g. control plane vs workloads) but it'd be worth getting that feedback from the experts.

> - Support installing OpenShift with 2 regular nodes and 1 arbiter node.

Reviewer 2: ```suggestion
- Support installing OpenShift with 2 master nodes and 1 arbiter node.
```

> - The arbiter node hardware requirement will be lower than regular nodes.

Reviewer 2: ```suggestion
- The arbiter node hardware requirements will be lower than regular nodes in both cost and performance.
```

> - The arbiter node hardware requirement will be lower than regular nodes.

Author: Good idea, adding it.

### Non-Goals

The below goals are not intended to be worked on now, but might be expansion
ideas for future features.

- Running the arbiter node offsite

> - Running the arbiter node as a VM local to the cluster

Reviewer: I think we should not exclude running the arbiter node on a hypervisor. There might be situations where this would actually be helpful. I think the key point is really not to run the arbiter on OCPVirt on the same cluster, as this spoils the idea of 3 node redundancy.

> - Running the arbiter node as a VM local to the cluster

Author: That makes sense to me, like a normal deployment, it shouldn't matter where the arbiter node lives. I'll update the wording.

- Having a single arbiter supporting multiple clusters

> - Moving from 2 + 1 to conventional 3 node cluster

Reviewer 2: ```suggestion
- Moving from 2 + 1 to a conventional 3 node cluster
```

Reviewer: Why are we stating this as a non-goal? The JIRA feature asks for this (see requirement number  6 in the description)

Author: Yeah you're right, I was thinking we would tackle that as it's own separate feature, with the other non-goals for future arbiter potential. I'll add it to goals.

> - Moving from 2 + 1 to conventional 3 node cluster

Reviewer: the etcd / ocp topology would not change to my understanding - it remains a 3 node cluster.

> - Moving from 2 + 1 to conventional 3 node cluster

Author: This is true, I was wondering what ability we had to change that behavior, but it seems it might be a bigger challenge. I'll remove this angle from the doc.

## Proposal

The main focus of the enhancement is to support edge deployments of individual
OpenShift HA clusters at scale, and to do so in a cost effective way. We are
proposing doing this through the incorporation of an arbiter node as a quasi
heterogenous control plane configuration. The arbiter will run the critical
components that help maintain an HA cluster, but other platform pods should not
be scheduled on the arbiter node. The arbiter node will be tainted to make sure
that only deployments that tolerate that taint are scheduled on the arbiter.

> Things that we are proposing of changing.

Reviewer 2: ```suggestion
Functionality that we are proposing to change:
```

- Adding a new topology to the [OCP/API Control Plane
  Topology](https://github.com/openshift/api/blob/69df64132c911e9eb0176e9697f451c13457e294/config/v1/types_infrastructure.go#L103)
  - This type of change should have an authoritative flag that indicates layout

>     of the control plane, this information would be valuable for operator

Reviewer 2: ```suggestion
    of the control plane which would be valuable for operator
```

    developers so no inference is required.
- We will add support to the OCP installer to provide a way of setting up the
  initial manifests and the ControlPlaneTopology field.
  - We will need to support a path for customers to indicate the desire for a 2

>     - 1 arbiter install configuration.

Reviewer 2: ```suggestion
    + 1 arbiter install configuration.
```

  - This will also be used to apply the taint to the machineset manifest.

> - Alter CEO to be aware of the arbiter node role type and allow it to treat it

Reviewer 2: ```suggestion
- Alter Cluster ETCD Operator (CEO) to be aware of the arbiter node role type and allow it to treat it
```

  as if it were a master node.
  - We will need CEO to create an ETCD member on the arbiter node to allow

>     quarom to happen

Reviewer 2: ```suggestion
    quarum to happen
```

- Update the tolerations of any critical or desired component that should be
  running on the arbiter node.

### Workflow Description

#### For Cloud Installs

> 1. User sits down at the computer.

Author: Lol at this point, it's part of the ceremony of an EP, but I'll remove it.

validate some of the desires in this proposal. In it's current version this
proposal is not exhaustive but will be filled out as we implement these goals.

We currently expect this feature to mainly be used by `baremetal` installs, or
specialized hardware that is built to take advantage of this type of
configuration. In the current design we make two paths for cloud and baremetal
installs in the installer. However, the cloud install is primarily for testing,
this might mean that we simplify the installer changes if we are the only ones
using cloud installs, since we can simply alter the manifests in the pipeline
with out needing to change the installer.

### Risks and Mitigations

The main risk in this enhancement is that because we are treating one of the
master nodes in a 3 node cluster as an arbiter, we are explicitly evicting
processes that would otherwise be a normal supported upstream configuration such
as a compact cluster. We run the risk of new components being critical to HA not
containing the proper tolerations for running on the arbiter node. One of the
mitigations we can take against that is to make sure we are testing installs and
updates.

> Another risk we run is customers using an arbiter node with improper disk speeds

Reviewer 3: I'm leaning towards suggesting that this risk be removed since we document [etcd disk best practices](https://docs.openshift.com/container-platform/4.16/scalability_and_performance/recommended-performance-scale-practices/recommended-etcd-practices.html) so they'd also apply here. With that in mind, I don't think this risk is specific to this enhancement. If anyone else wants to give this idea a :+1: then maybe it can be dropped otherwise it can be left as-is.

Author: I just want to capture on what we think the suggestions should be for mitigations. So I'll add this link and update the  wording here a bit to make it clear the direction we're going.

below that recommended for etcd, since etcd is sensitive to latency between
members, we should provide proper guidance so that the arbiter node doesn't
become a bottleneck for etcd.

### Drawbacks

A few drawbacks we have is that we will be creating a new variant of OpenShift
that implements a new unique way of doing HA for kubernetes. This does mean an

> increase in the test matrix and all together a different type of tests since

Reviewer 2: `...and all together a different type of tests since` 

since what?

Author: Ops, thanks! Fixed

## Open Questions [optional]

> 1. In the future it might be desired to add another master and convert to a

Author: Do you know if that's a hard rule? I have it in there as an open question to see if this is something we might want to consider for things such as this.

> 1. In the future it might be desired to add another master and convert to a

Author: Okay, I'll leave it as an open question but remove it as a desire to change the topology for now.

   compact cluster, do we want to support changing ControlPlaneTopology field

>    after the fact?

Reviewer: see above, this is already listed in the requirements and should be possible. Maybe not on initial release, but then on the next one.

Reviewer: Feature Requirement number 8 states: 
`It must be possible to explicitly schedule additional workload to the arbiter node. That is important for 3d party solutions (e.g. storage provider) which also have  quorum based mechanisms.`
The use case is e.g. ODF in replica-2 mode, where some ODF components also need 3 deployments (the ceph mon for example). 
How is this supported by this ER? I think we should mention / describe this case.

Author: I guess that's not clear here, the idea is that anyone that wishes to run on an arbiter node will need to just have the proper tolerations or directly target the node. I'll capture that in the doc more explicitly.

## Test Plan

WIP

- Running e2e test would be preferred but might prove to be tricky due to the
  asymmetry in the control plane

> - We need a strategy for validating install and test failures

Reviewer 2: We may need to modify a lot of the tests in origin to account for this new configuration or add a whole new test suite to accommodate it

Author: Good idea, I'll keep it as requiring a new lane for now, since I'm not sure if we would do that through a suite at this point.

## Graduation Criteria

### Dev Preview -> Tech Preview

- Ability to utilize the enhancement end to end
- End user documentation, relative API stability

> - Sufficient test coverage

Reviewer 2: How do we define sufficient here?

Reviewer: Whats the reason we need a Dev Preview? Could we start with a Tech Preview right away, to reduce potential time to market? We could extend TP phase if required.

Author: At this point that's fair, I think we have agreed on desiring a TechPreview first approach here. Changing it.

- Gather feedback from users rather than just developers
- Enumerate service level indicators (SLIs), expose SLIs as metrics
- Write symptoms-based alerts for the component(s)

### Tech Preview -> GA

- More testing (upgrade, downgrade, scale)
- Sufficient time for feedback
- Available by default
- Backhaul SLI telemetry
- Document SLOs for the component
- Conduct load testing
- User facing documentation created in
  [openshift-docs](https://github.com/openshift/openshift-docs/)

**For non-optional features moving to GA, the graduation criteria must include
end to end tests.**

### Removing a deprecated feature

N/A

## Upgrade / Downgrade Strategy

WIP

## Version Skew Strategy

N/A

## Operational Aspects of API Extensions

WIP

## Support Procedures

WIP

## Alternatives

We originally had tried using the pre-existing features in OCP, such as setting
a node as NoSchedule to avoid customer workloads going on the arbiter node.
While this whole worked as expected, the problem we faced is that the desire is

> to use a very lower powered and cheap device as the arbiter, this method would

Reviewer 2: ```suggestion
to use a device that is lower power and is cheaper as the arbiter. This method would
```

> still run a lot of the overhead on the arbiter node.

Reviewer 2: ```suggestion
still run most of the OCP overhead on the arbiter node.
```

---
title: arbiter-clusters
authors:
  - "@eggfoobar"
reviewers:
  - "@tjungblu"
  - "@patrickdillon"
  - "@racedo"
  - "@deads2k"
  - "@jerpeter1"

> approvers:

Author: Ah, yup, updating this, thanks!

	// credential information for communicating with management
	// controllers on hosts.
	SecretFiles []*asset.File

	// NetworkConfigSecretFiles is used by the baremetal platform to
	// store the networking configuration per host
	NetworkConfigSecretFiles []*asset.File

	// HostFiles is the list of baremetal hosts provided in the
	// installer configuration.
	HostFiles []*asset.File
}
```

### ETCD Operator Change

CEO(cluster-etcd-operator) will need to be made aware of the new arbiter role.
There are some challenges posed by this since CEO uses the node informer for
`master` in a lot of their controllers to manage the etcd cluster. We also have
a challenge in that the `node-role` labels are just empty keys, meaning that a
label selector like `node-role in (master,arbiter)` will not work for
`node-role` since we always just match on the key's existence. Furthermore,
labels selectors by design do not support an `or` operation on keys
([see](https://github.com/kubernetes/kubernetes/issues/90549#issuecomment-620625847)).
We will need to wrap the informer and lister interfaces to allow much of the
logic not to change for CEO, this does also offer a benefit since it will be

> easier to FeatureGate during our TechPreview phase.

Reviewer 4: I don't think we're going to feature gate a "joint informer" :) that doesn't seem worth the hassle

Author: Lol that's fair, I'll clean up that line.

When the feature is turned on, the custom informer and lister will be used to
watch nodes that have `node-role.kubernetes.io/master` or
`node-role.kubernetes.io/arbiter` labels.

> `starter.go`

Author: Sounds good, removing.

validate some of the desires in this proposal. In it's current version this
proposal is not exhaustive but will be filled out as we implement these goals.

We currently expect this feature to mainly be used by `baremetal` installs, or
specialized hardware that is built to take advantage of this type of
configuration. In the current design we make two paths for cloud and baremetal
installs in the installer. However, the cloud install is primarily for testing,
this might mean that we simplify the installer changes if we are the only ones
using cloud installs, since we can simply alter the manifests in the pipeline
with out needing to change the installer.

### Risks and Mitigations

The main risk in this enhancement is that because we are treating one of the
master nodes in a 3 node cluster as an arbiter, we are explicitly evicting
processes that would otherwise be a normal supported upstream configuration such
as a compact cluster. We run the risk of new components being critical to HA not
containing the proper tolerations for running on the arbiter node. One of the
mitigations we can take against that is to make sure we are testing installs and
updates.

> Another risk we run is customers using an arbiter node with improper disk speeds

Reviewer 4: I think there's a much higher risk that the network latency is impacting etcd negatively than the disk speed. 

See also https://github.com/etcd-io/etcd/issues/14501

Author: Ah, you're right, updated and thanks for the link, added it :)

---
title: arbiter-clusters
authors:
  - "@eggfoobar"
reviewers:
  - "@tjungblu"
  - "@patrickdillon"
  - "@racedo"
  - "@deads2k"
  - "@jerpeter1"
  - "@sjenning"
  - "@yuqi-zhang"
  - "@zaneb"
  - "@rphillips"
approvers:
  - "@tjungblu"
  - "@patrickdillon"
  - "@racedo"
  - "@jerpeter1"
  - "@deads2k"
  - "@sjenning"
  - "@yuqi-zhang"
  - "@zaneb"
api-approvers:
  - "@JoelSpeed"
creation-date: 2024-08-27
last-updated: 2024-10-16
tracking-link:
  - https://issues.redhat.com/browse/OCPEDGE-1191
see-also: []
replaces: []
superseded-by: []
---

# Support 2 Node + 1 Arbiter Node HA Cluster

## Summary

This enhancement describes an ability to install OpenShift with a control plane
that consists of at minimum 2 normal sized nodes, and at least 1 node that can
be less powerful than the recommended node size. This 1 arbiter node will only
be running critical components for maintaining HA to allow the arbiter node size
to be as small and as low cost as possible with in reason.

## Motivation

Customers at the edge are requiring a more economical solution for HA
deployments at the edge. They can support running 2 node clusters for redundancy

> but would like the option to deploy a lower cost node as an arbiter to supply

Reviewer 5: It would also be worth considering having an arm64 node for the arbiter to lower costs and deploy the cluster with the multi-arch payload.

Reviewer 5: is the arbiter node considered strictly control plane?

Reviewer 6: AIUI we know the specific hardware we are targeting here (it has the low-powered node built-in).

  full master node.
- HA for a 2+1 arbiter node should match the HA guarantees of a 3 Node Cluster
  deployment.
- Support installing OpenShift with 2 master nodes and 1 arbiter node.
- Add a new ControlPlaneTopology and InfrastructureTopology in the
  Infrastructure API
- The arbiter node hardware requirements will be lower than regular nodes in
  both cost and performance. Customers can use devices on the market from OEMs
  like Dell that supply an all in one unit with 2 compute and 1 lower powered
  compute for this deployment scenario.

### Non-Goals

The below goals are not intended to be worked on now, but might be expansion
ideas for future features.

- Running the arbiter node offsite.
- Running a virtualized arbiter node on the same cluster.
- Having a single arbiter supporting multiple clusters.
- Moving from 2 + 1 to a conventional 3 node cluster

## Proposal

The main focus of the enhancement is to support edge deployments of individual
OpenShift HA clusters at scale, and to do so in a cost effective way. We are
proposing doing this through the creation of a new node role type called an
`node-role.kubernetes.io/arbiter` node as a heterogenous quasi control plane

> configuration. The arbiter will run the critical components that help maintain

Author: That's true, the wording is poor here since I was thinking of things like MCD here as critical for our purposes, but it's not strictly needed for HA. I'll update it to be clear.

an HA cluster, but other platform pods should not be scheduled on the arbiter
node. The arbiter node will be tainted to make sure that only deployments that
tolerate that taint are scheduled on the arbiter.

We think creating this new node role would be the best approach for this
feature. Having a new node role type allows our bootstrap flow to leverage the
existing MCO and Installer systems to supply specific Ignition files for a
smaller footprint device. Existing deployments will essentially be blind to this
node role unless they need to be made aware of it, but critical workloads that
allow all taints can still run on the node, thus making it easier for smaller
arbiter nodes to be deployed with out higher memory and cpu requirements.

> Components that we are proposing to change:

Reviewer 7: HCP to include ban on installing on this topology.  Probably also ban installing on single node.

> Components that we are proposing to change:

Reviewer 7: OLM to do some work fanning out for an infrastructure annotation.

Author: Is your thinking here to add support to OLM to filter components based off of topology type?

> | Component                                                  | Change                                                                                                          |

Reviewer 8: Assisted installer needs to be updated, as it expects 3 nodes in non-SNO mode and uses one of masters for bootstrap. I suppose it should prefer the arbiter

Reviewer 9: I guess also Assisted Service will require to be updated as well in order to support a new role for a less powerful hardware (ie, for the [validations requirements](https://github.com/openshift/installer/blob/8032a54a41008ddf372a810ab6ff9a1b95880251/data/data/agent/files/usr/local/share/assisted-service/assisted-service.env.template#L12) configured in ABI)

Author: Great call out, thanks guys, adding them in here for now as a high level addition, But I'll do some research to make sure I capture detail on the units that need to be changed.

| ---------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------- |
| [Infrastructure API](#infrastructure-api)                  | Add `HighlyAvailableArbiter` as a new value for `ControlPlaneTopology` and `InfrastructureTopology`             |
| [Installer](#installer-changes)                            | Update install config API to Support arbiter machine types                                                      |
| [MCO](#mco-changes)                                        | Update validation hook to support arbiter role and add bootstrapping configurations needed for arbiter machines |
| [Kubernetes](#kubernetes-change)                           | Update allowed `well_known_openshift_labels.go` to include new node role as an upstream carry                   |

> | [ETCD Operator](#etcd-operator-change)                     | Update operator to deploy operands on both `master` and `arbiter` node roles                                    |

Reviewer 7: If the arbiter must not be the leader, then that's an additional feature that needs to be listed and designed.

Author: After some discussion on what we want to happen here. We're okay with the Arbiter node etcd member becoming leader. We're under the impression that etcd will be as fast as it's slowest member, then this might be fine for now. I'll document this as an open question, and plan to revisit this as it's own potential EP, wdyt?

@tjungblu Any concerns with this assumption on my part?

Reviewer 4: there is no functionality that can prevent it becoming a leader. I'm mostly concerned about the performance impact reported here: https://github.com/etcd-io/etcd/issues/14501

There are some upstream k8s PRs in the works that enable grpc "circuit breakers", so this will get resolved at some point in the future

| [library-go](#library-go-change)                           | Update the underlying static pod controller to deploy static pods to `arbiter` node roles                       |
| [Authentication Operator](#authentication-operator-change) | Update operator to accept minimum 2 kube api servers when `ControlPlaneTopology` is `HighlyAvailableArbiter`    |

### Infrastructure API

We need an authoritative flag to make it easy for components and tools to
identify if they are running against an `arbiter` backed HA cluster. We are

> proposing adding a new `ControlPlaneTopology` and `InfrastructureTopology` key

Author: That's fair, I wasn't sure if there would be a future need for distinction if users were to add workers to these clusters. Not something that's supported now, given the desire here is low cost deployment.

field mutable to allow a cluster to transition between topologies. However, that
conversation is beyond the scope of this enhancement, and requires its own
dedicated enhancement. We will keep this field as static and not support
topology migrations.

### Installer Changes

We will need to update the installer to have awareness of the explicit intent to
setup the cluster with an arbiter node. Adding a new machine type similar to the
flow for `master` and `workers` machines. As an example if the intent is to
generate 1 `arbiter` machine and 2 `master` machines, the user specifies a
`MachinePool` `installConfig.arbiterNode` field with the desired replicas.

The `arbiterNode` will be a `MachinePool` object that will enforce at least 1
replica, and at least 2 replicas for the `controlPlane`. When no `arbiterNode`
field is supplied regular flow will validate. When `arbiterNode` is supplied 2
replicas will be valid for for `controlPlane` as long as an `arbiterNode` is
specified.

`installConfig.yaml`

```yaml
apiVersion: v1
baseDomain: devcluster.openshift.com
compute:
  - architecture: amd64
    hyperthreading: Enabled
    name: worker
    platform: {}
    replicas: 0

> arbiterNode:

Reviewer 6: A simpler alternative - that would support baremetal clusters only - is to assume that the arbiter is part of the control plane, and identify its existence only by applying the `arbiter` role to one of the hosts in `platform.baremetal.hosts`.
This would result in the _installer_ (rather than a MachineSet) creating the arbiter Node with a taint and a separate pointer ignition.

Author: Originally I was leaning that way, but we also did have some testing and cloud deployment desires, that I switched the flow to follow the typical install flow for master, so now it's a full control plane machine at install. I'll update that step by step to correctly reflect that, my mistake on the `MachineSet` wording.

another check to this switch statement to allow minimum 2 replicas for
`HighlyAvailableArbiter` topologies.

### Workflow Description

#### For Cloud Installs

1. The user creates an `install-config.yaml`.
2. The user defines the `installConfig.controlPlane` field with `2` replicas.
3. The user then enters the new field `installConfig.arbiterNode` defines the
   arbiter node.
4. The user generates the manifests with this install config via the
   `openshift-install create manifests`
5. The installer creates a new `arbiter` MachineSet with a replica of `1` and
   checks the control plane replicas is `2` or higher
6. The installer sets the control plane and infrastructure topology to
   `HighlyAvailableArbiter`
7. With the object `arbiterNode` in the install config, the installer creates
   the machine object with the correct node labels and machine config resources
   for ignition bootstrap files.
8. The installer applies the taint to the arbiter MachineSet
9. The user can make any alterations to the node machine type to use less
   powerful machines.
10. The user then begins the install via `openshift-install create cluster`

> #### For Baremetal Installs

Reviewer 6: We should also be thinking about how to add support for this in the assisted installer, ZTP, and the agent-based installer.
One of our goals at the moment is to close all of the gaps between metal IPI and ABI, with a view to incorporating the latter into the former. So new gaps will not be welcomed :slightly_smiling_face:

Author: Fully agree, I'll add a section on that path as well, I don't have a clear idea on that due to unfamiliarity with assisted. I'll document our desire here and update early next week after some research.

1. The user creates an `install-config.yaml` like normal.
2. The user defines the `installConfig.controlPlane` field with `2` replicas.
3. The user then enters the new field `installConfig.arbiterNode` defines the
   arbiter node.
4. The user then enters the machine information for
   `installConfig.controlPlane.platform.baremetal` and
   `installConfig.arbiterNode.platform.baremetal`

> 5. The installer creates a new `arbiter` MachineSet with a replica of `1` and

Reviewer 6: So the arbiter node will be created like a worker, i.e. provisioned by a MachineSet in the cluster after the control plane comes up? (As opposed to like the control plane nodes, which are created directly by the bootstrap and do not have a MachineSet.)
We may have more work to do around labelling BaremetalHosts and filtering by label in the MachineSet so that the correct BaremetalHost is selected as the arbiter node.

Author: Ah, nope, this is a mistake in my original POC wording, thanks for pointing that out. The arbiter node follows the same install path as a master, so we create 2 master machines and 1 arbiter machine. They come up and work normally, except for one of the masters being only an arbiter.

Regarding BaremetalHost, I have not POCed that flow just yet, but I was also thinking labeling a host explicitly would be the path there. I'll flush out that flow a bit clearer here.

#### Hypershift / Hosted Control Planes

At the time being there is no impact on Hypershift since this edge deployment
will require running the control plane.

#### Standalone Clusters

This change is relevant to standalone deployments of OpenShift at the edge or
datacenters. This enhancement specifically deals with this type of deployment.

#### Single-node Deployments or MicroShift

This change does not effect Single Node or MicroShift.

### Implementation Details/Notes/Constraints

Currently there are some behavior unknowns, we will need to test out our POC to
validate some of the desires in this proposal. In it's current version this
proposal is not exhaustive but will be filled out as we implement these goals.

We currently expect this feature to mainly be used by `baremetal` installs, or
specialized hardware that is built to take advantage of this type of
configuration. In the current design we make two paths for cloud and baremetal
installs in the installer. However, the cloud install is primarily for testing,
this might mean that we simplify the installer changes if we are the only ones
using cloud installs, since we can simply alter the manifests in the pipeline

> with out needing to change the installer.

Reviewer 6: I would seriously question whether the work saved in manually setting up tests for etcd justifies the extra work in supporting this on cloud platforms.
Not only are cloud platforms an extra thing to support, but installing the arbiter node via a MachineSet in a platform-independent way complicates the baremetal implementation, and the Assisted/ABI implementations as well.

Author: I understand the concern, I'll try to capture this better in the documentation, but there is value for cloud deployment, especially as a TechPreview iteration of this feature. Currently there are some unknowns around where users are with the hardware they will be using, so having a cloud environment would be easier for them to deploy and spin the tires on it as well as make testing and validation a bit easier for us.

I went ahead and updated the doc to reflect the fact that the arbiter node is being installed in the same way as the master nodes, would that help smooth out your concerns?

Author: We realigned the focus to just baremetal for the time being, the doc has been updated to focus only on baremetal.

In the current POC, a cluster was created on AWS with 2 masters running
`m6a.4xlarge` and 1 arbiter running a `m6a.large` instance. At idle the node is
consuming `1.41 GiB / 7.57 GiB	0.135 cores / 2 cores`. The cluster installs and
operates with the following containers running on the arbiter node. This seems
very promising, but we will need to identify any other resources we might want
to run or remove from this list.

```

> "openshift-cluster-csi-drivers/aws-ebs-csi-driver-node-qvvqc"

Reviewer 8: probably not needed - no required component provisions EBS volumes

Author: Thanks for the info, adding it to a table with this list to help keep track.

"openshift-cluster-node-tuning-operator/tuned-q7f98"
"openshift-dns/node-resolver-zcp7j"
"openshift-etcd/etcd-guard-ip-10-0-21-103.us-west-2.compute.internal"
"openshift-etcd/etcd-ip-10-0-21-103.us-west-2.compute.internal"

> "openshift-image-registry/node-ca-4cxgp"

Reviewer 8: not needed - no pods use image registry

A couple of risks we run is customers using an arbiter node with improper disk
speeds below that recommended for etcd, or a bigger problem being network
latency. Since etcd is sensitive to latency between members (see:
[etcd-io/etcd#14501](https://github.com/etcd-io/etcd/issues/14501)), we should
provide proper guidance so that the arbiter node must meet minimum requirements
for ETCD to function properly, in disk and network speeds. ([ETCD
Recommendations](https://docs.openshift.com/container-platform/4.16/scalability_and_performance/recommended-performance-scale-practices/recommended-etcd-practices.html))

### Drawbacks

A few drawbacks we have is that we will be creating a new variant of OpenShift
that implements a new unique way of doing HA for kubernetes. This does mean an
increase in the test matrix and all together a different type of tests since the
addition of an arbiter node will require different validation scenarios for
failover.

## Open Questions [optional]

1. In the future it might be desired to add another master and convert to a
   compact cluster, do we want to support changing ControlPlaneTopology field
   after the fact?

> ## Test Plan

Reviewer 7: OLM+whoever to provide a way to make the arbiter configuration easy to test.

cc @joelanford

- We will create a CI lane to validate install and fail over scenarios such as
  loosing a master or swaping out an arbiter node.
- Create complimentary lanes for serial/techpreview/no-capabilities.
- Create a lane for or some method for validating layered products.
- CI lane for e2e conformance testing, tests that explicitly test 3 node masters
  will need to be altered to accommodate the different topology.
- We will add e2e tests to specifically test out the expectations in this type
  of deployment.

>   - Create tests to validate correct pods are running on the `arbiter`.

Reviewer 7: Also need to test that no incorrect pods are running on the arbiter.

Author: Sounds good adding to doc.

  - Create tests to validate routing and disruptions are with in expectations.

## Graduation Criteria

### Dev Preview -> Tech Preview

N/A

### Tech Preview -> GA

- More testing (upgrade, downgrade, scale)
- E2E test additions for this feature
- Sufficient time for feedback
- Available by default
- Backhaul SLI telemetry
- Document SLOs for the component
- Conduct load testing
- User facing documentation created in
  [openshift-docs](https://github.com/openshift/openshift-docs/)

Before going GA we need to have a proper idea of the role that the
[control-plane machine-set

> operator](https://github.com/openshift/cluster-control-plane-machine-set-operator/tree/main/docs/user#overview)

Reviewer 6: We don't use ControlPlaneMachineSetOperator on `platform:baremetal`, so you only have to deal with this for clouds.

- As an OpenShift cluster admin I want to be able to allow deployments with
  proper tolerations or explicitly defined node in the `spec` to be able to be
  scheduled on the arbiter node.

### Goals

- Provide a new arbiter node role type that achieves HA but does not act as a
  full master node.
- HA for a 2+1 arbiter node should match the HA guarantees of a 3 Node Cluster
  deployment.
- Support installing OpenShift with 2 master nodes and 1 arbiter node.
- Add a new ControlPlaneTopology and InfrastructureTopology in the
  Infrastructure API
- The arbiter node hardware requirements will be lower than regular nodes in
  both cost and performance. Customers can use devices on the market from OEMs
  like Dell that supply an all in one unit with 2 compute and 1 lower powered
  compute for this deployment scenario.

### Non-Goals

The below goals are not intended to be worked on now, but might be expansion
ideas for future features.

- Running the arbiter node offsite.
- Running a virtualized arbiter node on the same cluster.
- Having a single arbiter supporting multiple clusters.
- Moving from 2 + 1 to a conventional 3 node cluster

> 

Reviewer 10: - Supporting workers of type `arbiter`?

Author: No, thankfully, that's not in plan for this, the idea is that the Arbiter will function as a smaller less powerful Master, so there should never be an Arbiter that is a worker

Author: Added some more explicit wording around this so that's captured

OpenShift HA clusters at scale, and to do so in a cost effective way. We are
proposing doing this through the creation of a new node role type called an
`node-role.kubernetes.io/arbiter` node as a heterogenous quasi control plane
configuration. The arbiter will run the minimum components that help maintain an
HA cluster, things like MCD, monitoring and networking. Other platform pods
should not be scheduled on the arbiter node. The arbiter node will be tainted to
make sure that only deployments that tolerate that taint are scheduled on the
arbiter.

We think creating this new node role would be the best approach for this
feature. Having a new node role type allows our bootstrap flow to leverage the
existing MCO and Installer systems to supply specific Ignition files for a
smaller footprint device. Existing deployments will essentially be blind to this
node role unless they need to be made aware of it, but critical workloads that
allow all taints can still run on the node, thus making it easier for smaller
arbiter nodes to be deployed with out higher memory and cpu requirements.

Components that we are proposing to change:

> | Component                                                     | Change                                                                                                          |

Author: Thanks Patrick, the control plane machine set operator should not be in play for this feature at the time being, since our focus is bare metal for support, and if we do end up doing cloud installs for CI validation of core components, we will be making that operator `Inactive` but will need to revisit it for any cloud support.

> | Component                                                     | Change                                                                                                          |

Author: It doesn't seem like the baremetal operator will need changes from my preliminary POC.

| ------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------- |
| [Infrastructure API](#infrastructure-api)                     | Add `HighlyAvailableArbiter` as a new value for `ControlPlaneTopology` and `InfrastructureTopology`             |

> | [Installer](#installer-changes)                               | Update install config API to Support arbiter machine types                                                      |

Author: I'm still in the process of deploying a baremetal cluster with these changes, but ideally since this is going techpreview, would love an easy way for customers to try it out, so I'll hope to have an IPI flow as well for ease of deployment.

After some conversations around what CPMSO does, since we are focusing on baremetal for the time being, we can punt that question for any future conversations of cloud deployments. 

I do have a WIP PR here https://github.com/openshift/installer/pull/9159 but that will probably change with the focus on baremetal flow.

`custom-machine-config-pool-selector-validatingadmissionpolicy.yaml`

```
          has(object.spec.machineConfigSelector.matchLabels) &&
          (
            (object.spec.machineConfigSelector.matchLabels["machineconfiguration.openshift.io/role"] == "master")
            ||
            (object.spec.machineConfigSelector.matchLabels["machineconfiguration.openshift.io/role"] == "worker")
            ||
            (object.spec.machineConfigSelector.matchLabels["machineconfiguration.openshift.io/role"] == "arbiter")
          )
```

We will also need to update the go logic in the renderer to allow the `arbiter`
role.

`render.go`

```go
// GenerateMachineConfigsForRole creates MachineConfigs for the role provided
func GenerateMachineConfigsForRole(config *RenderConfig, role, templateDir string) ([]*mcfgv1.MachineConfig, error) {
	rolePath := role
	//nolint:goconst
	if role != "worker" && role != "master" && role != "arbiter" {
		// custom pools are only allowed to be worker's children
		// and can reuse the worker templates
		rolePath = "worker"
	}
...
```

We then need to add new rendered initial files for the arbiter role so

> bootstrapping occurs correctly. When MCO stands up the Machine Config Server

Reviewer 11: As a note this also creates a set of arbiter-user-data and arbiter-user-data-managed secrets in the mapi namespace. I don't suppose we'd want to scale additional arbiter nodes post-install though, so those would just be unused (which is probably fine?) but the user would theoretically be able to leverage them to scale additional arbiter nodes.

Also to check, do we foresee anyone adding arbiter nodes post-install? Maybe we can just have the bootstrap MCS server arbiter and block the in-cluster MCS from doing so.

Author: At the time being there is no need to scale the arbiter node, but there is no reason why we can't do that in the future for small HA deployments. In the installer changes I left the replica field to allow us to revisit this in the future, we can put in a validation check for now to enforce 1 arbiter. However, I can imagine a future scenario where we configure a 2 Arbiter 1 Master.

today, we supply initial configs from the

> [template/<master/worker>/00-<master/worker>](https://github.com/openshift/machine-config-operator/tree/master/templates)

Reviewer 11: One thing here is I'd like to make it such that any cluster that isn't using arbiter nodes to not have an arbiter pool. As it stands I think the proposal will make it such that we always create an arbiter pool, just empty and with no nodes. Since it almost looks like a custom pool, I'd prefer we don't confuse users into thinking we generated a new pool for them.

Author: That's a great call out, I'll add that wording to make sure we only create this pool when the arbiter is specified

that are rendered in the [template
controller](https://github.com/openshift/machine-config-operator/blob/7c5ae75515fe373e3eecee711c18b07d01b5e759/pkg/controller/template/render.go#L102).

> We will need to add the appropriate folder with the files relevant to the

Author: Yeah agreed, I'll reevaluate this change and make some notes for MCO to help noodle some ideas on de-duplication of files for GA

> We will need to add the appropriate folder with the files relevant to the

Author: left a node on road to GA to make sure we have a solution for this before hitting GA.

---
title: arbiter-clusters
authors:
  - "@eggfoobar"
reviewers:

>   - "@tjungblu"

Reviewer 4: lgtm

  - "@patrickdillon"

>   - "@racedo"

Reviewer 12: lgtm

We need an authoritative flag to make it easy for components and tools to
identify if they are running against an `arbiter` backed HA cluster. We are
proposing adding a new `ControlPlaneTopology` key called
`HighlyAvailableArbiter` to denote that the cluster is installed with an
arbiter.

Currently this field will not change, once a cluster is installed as an arbiter
HA cluster, it will remain that way. There has been discussion on making this
field mutable to allow a cluster to transition between topologies. However, that
conversation is beyond the scope of this enhancement, and requires its own
dedicated enhancement. We will keep this field as static and not support
topology migrations.

### Installer Changes

We will need to update the installer to have awareness of the explicit intent to
setup the cluster with an arbiter node. Adding a new machine type similar to the
flow for `master` and `workers` machines. As an example if the intent is to
generate 1 `arbiter` machine and 2 `master` machines, the user specifies a
`MachinePool` for `installConfig.arbiter` field with the desired replicas.

The `arbiter` will be a `MachinePool` object that will enforce at least 1

> replica, and at least 2 replicas for the `controlPlane`. When no `arbiter` field

Author: Yup, this is correct, I'll update the doc to make that explicit

is supplied regular flow will validate. When `arbiter` is supplied 2 replicas
will be valid for for `controlPlane` as long as an `arbiter` is specified.

Some validation we will enforce:

1. Only on BareMetal installs.
2. Minimum 1 Arbiter replica if Arbiter is defined.
3. Minimum 2 Master replica when Arbiter is defined.

`installConfig.yaml`

```yaml
apiVersion: v1
baseDomain: devcluster.openshift.com
compute:
  - architecture: amd64
    hyperthreading: Enabled
    name: worker
    platform: {}
    replicas: 0
arbiter:
  architecture: amd64
  hyperthreading: Enabled
  replicas: 1
  name: arbiter
  platform:
    baremetal: {}
controlPlane:
  architecture: amd64
  hyperthreading: Enabled
  name: master
  platform:
    baremetal: {}
  replicas: 2
platform:
  baremetal:
    ...
    hosts:
      - name: cluster-master-0
        role: master
        ...
      - name: cluster-master-1
        role: master
        ...
      - name: cluster-arbiter-0

>         role: arbiter

Reviewer 6: For Assisted/Agent-based installs where the role is not required, are there any heuristics we should apply for choosing the arbiter node? Presumably assisted-service should choose the smallest node (by RAM/CPU) when asked to allocate automatically?

Author: Yeah, that's good point, the smallest node would be the appropriate path forward, I'll capture this in the doc.

arbiter topology is explicitly chosen during install. The MCP and resources
described here in the MCO changes should only be applied if the arbiter topology
is explicitly turned on.

`/templates/arbiter/**/kubelet.service.yaml`

```
        --node-labels=node-role.kubernetes.io/arbiter,node.openshift.io/os_id=${ID} \
...
        --register-with-taints=node-role.kubernetes.io/arbiter=:NoSchedule \
```

The initial TechPreview of this feature will copy over the master configurations
with slight alterations for the `arbiter`. However, for GA we will need to
create a more appropriate flow for this so we're not duplicating so much.

### Kubernetes Change

Our fork of kubernetes contains a small addition for custom node roles thats
used to circumvent the [kubelet flag
validation](https://github.com/openshift/kubernetes/blob/6c76c890616c214538d2b5d664ccfcb4f5e460bd/cmd/kubelet/app/options/options.go#L158)
preventing a node from containing roles that use the prefix
`node-role.kubernetes.io`. We will need modify
[well_known_openshift_labels.go](https://github.com/openshift/kubernetes/blob/master/staging/src/k8s.io/kubelet/pkg/apis/well_known_openshift_labels.go)

> to include the new `node-role.kubernetes.io/arbiter`. This will be an upstream

Reviewer 6: s/upstream/downstream/

    nodes = append(nodes, arbiternodes...)
}
...

```

### Authentication Operator Change

The authentication operator runs a [readiness check
controller](https://github.com/openshift/cluster-authentication-operator/blob/2d71f164af3f3e9c84eb40669d330df2871956f5/pkg/controllers/readiness/unsupported_override.go#L58-L70)
that enforces for all non `SingleReplicaTopologyMode` topologies or an
explicitly override, there must be 3 or more `kube-apiservers`. We need to add
another check to this switch statement to allow minimum 2 replicas for
`HighlyAvailableArbiter` topologies.

### Hosted Control Plane Change

We will not support running this type of topology for hosted control plane
installations. We need to update HyperShift to check for the
`HighlyAvailableArbiter` control plane topology and prevent HCP installations on
Arbiter clusters. This will need to be done at the `hypershift` cli level as
well as any bootstrap component that can be created outside of the CLI flow.

### Alternative Install FLow Change

We currently have a few different options for different needs when installing
OCP that need to also be updated. Work done in the installer should be reflected

> on the Assisted Installer, Assisted Service and ZTP.

Author: Thanks so much for the insight here, I like the option you proposed more, I'll go ahead and add at to the doc.

#### During Install

1. The CEO will watch for new masters and the arbiter role
2. CEO will create the operand for the etcd deployments that have tolerations
   for the arbiter
3. Operators that have tolerations for the arbiter should be scheduled on the
   node
4. The install should proceed as normal

### API Extensions

#### Installer API Change

The `installConfig` will include a `installConfig.arbiter` object for a
`MachinePool` to configure arbiter infra structure.

#### OCP Config API Change

The Infrastructure config fields for `ControlPlaneTopology` will support the new
value of `HighlyAvailableArbiter`

`InfrastructureTopology` will be left as `HighlyAvailable` in this situation
since the existing rule is we only highlight when we are dealing with
`SingleReplica` workers. In this instance, at minimum 2 masters will be able to
schedule worker workloads, thus follow the same pattern as a 3 Node cluster.
This makes the most sense for now since having `InfrastructureTopology` denote a
halfway point between single and three might be a moot distinction for workers.
However, we might need to revisit this as a product team might assume a

> `InfrastructureTopology: HighlyAvailable` == `3 Workers`.

Reviewer 6: This shouldn't be a problem as a cluster with 3 control plane nodes and 2 workers would still have `InfrastructureTopology: HighlyAvailable`.

Out of curiosity, is it implied and/or required that clusters with an arbiter node are also created without workers (i.e. mastersSchedulable)?

Reviewer 6: OK, so we don't need to have a validation that mastersSchedulable is enabled, and the default for it should continue to be based only on whether the number of workers is <2 :+1:

### Drawbacks

A few drawbacks we have is that we will be creating a new variant of OpenShift
that implements a new unique way of doing HA for kubernetes. This does mean an
increase in the test matrix and all together a different type of tests since the
addition of an arbiter node will require different validation scenarios for
failover.

## Open Questions [optional]

1. In the future it might be desired to add another master and convert to a
   compact cluster, do we want to support changing ControlPlaneTopology field
   after the fact?

2. Do we need to modify OLM to filter out deployments based on topology
   information?

3. Since the focus of this feature is BareMetal for GA, we don't need to worry
   about the [control-plane machine-set
   operator](https://github.com/openshift/cluster-control-plane-machine-set-operator/tree/main/docs/user#overview).
   However, we will need to revisit this so we a proper idea of the role that it
   plays in this type of configuration.

## Test Plan

- We will create a CI lane to validate install and fail over scenarios such as

>   loosing a master or swaping out an arbiter node.

Reviewer 6: ```suggestion
  losing a master or swapping out an arbiter node.
```

	rolePath := role
	//nolint:goconst
	if role != "worker" && role != "master" && role != "arbiter" {
		// custom pools are only allowed to be worker's children
		// and can reuse the worker templates
		rolePath = "worker"
	}
...
```

We then need to add new rendered initial files for the arbiter role so
bootstrapping occurs correctly. When MCO stands up the Machine Config Server
today, we supply initial configs from the
[template/<master/worker>/00-<master/worker>](https://github.com/openshift/machine-config-operator/tree/master/templates)
that are rendered in the [template
controller](https://github.com/openshift/machine-config-operator/blob/7c5ae75515fe373e3eecee711c18b07d01b5e759/pkg/controller/template/render.go#L102).
We will need to add the appropriate folder with the files relevant to the
arbiter, which in this iteration will be similar to master, except the kubelet
service will contain different initial
[node-roles](https://github.com/openshift/machine-config-operator/blob/7c5ae75515fe373e3eecee711c18b07d01b5e759/templates/master/01-master-kubelet/_base/units/kubelet.service.yaml#L31)
and

> [taints](https://github.com/openshift/machine-config-operator/blob/7c5ae75515fe373e3eecee711c18b07d01b5e759/templates/master/01-master-kubelet/_base/units/kubelet.service.yaml#L43).

Reviewer 6: @eggfoobar did we consider reusing the existing control-plane role tag and only adding an extra taint (which only the etcd components would tolerate) via the MachineSet? It seems like that would vastly reduce the complexity we'd have to maintain.

Author: Hey @zaneb, sorry for the late reply, didn’t see the notification for this until today.

That’s a great question, In short we hadn’t tried a different Taint but we did originally try `master:NoSchedule`, that worked for the most part and we did consider using a different Taint instead but a couple of design considerations pulled us away from it.

- We needed a quasi control-plane node that supports HA but should be smaller and treated differently in terms of what runs on it and how we configure it. This node will definitely be smaller with different characteristics and in the future this might even be an ARM node or in a different location.
- This has the high potential to be disruptive to components and the platform so we need to be explicit about the state of the cluster and allow component owners the signaling to decide how their component handles this deployment type.

Due to the uniqueness of the control plane, being a master node with a different taint poses different challenges. A big one is that things become implicit and a bit ambiguous. We currently have a few operators that tolerate all taints on control plane nodes, etcd being one of them. Updating those components would be easy enough so we can exclude them from the arbiter node but they would need to still check some authoritative flag that we are indeed in an arbiter deployment for replica modifications. In this case I think we continue to keep the ControlPlane API change. So you can probably get away with having taints be your enforcement for what runs on the arbiter, but with the API change we’re taking just a different road to the same path. The ambiguity comes from the customers side, we run the risk of components installed later running on the arbiter that shouldn't because they tolerate all taints on masters or not run at all. The cluster will says 3 control-plane nodes, so I can see where that might be confusing and the customer and they might just remove the taint from googling for solutions. The upside here is that we make fewer installer changes but we just gotta make sure the authoritative flag controlPlane API is set.

However, once we started to think about configuration aspect, having a machine that inherits from master is something we don’t support due to stability concerns, plus high probability that we would want to alter the same fields as master node configs. So the more we thought about it, the more we came back to the intent, we needed a quasi control plane node and the correct way to represent that seemed to be an explicit node role that was clear in its purpose. 

So the gamble is we pay a little complexity upfront or we would pay it later when customers already had workloads running and tough decisions would mean tough conversations. If I recall correctly those were the main ideas running around my head a few months back when we were experimenting with different implementations.

Reviewer 13: The other observation I would make is that since I've been part of OpenShift, there has been an assumption that control-plane nodes are expected to be homogenous. We softly enforce this for cloud-based IPI installs by allowing a single instance type to be provisioned for all control-plane nodes.

I know we would support breaking this assumption for UPI clusters, and it might also be possible to have mismatched control plane nodes in a baremetal IPI install, but I'm not aware of any testing that verifies that mismatched control plane sizes behave as expected.

Keeping this as a separate node role helps us make this boundary more explicit. This node is expected to be different, and will thus also be tested as such.

on the Assisted Installer Service, Agent Based Installer and ZTP.

In ABI (Agent Based Installer) the bootstrap node, which runs the
`assisted-service` as well as the `in-place bootstrap`, is required to be a
control plane node. This should not run on the arbiter node since it will have
less compute and thus we would prefer it to run on one of the larger control
plane nodes.

In both ABI and Assisted, the node that runs the `in-place bootstrap` is the
last to be rebooted to join the cluster. We would require this to be the one of
the large control plane node, and since 2 etcd replicas are all that is required
to add the third node safely, a 1 control plane + 1 arbiter cluster satisfies
this requirement. This also would mean less major code changes.

### OLM Filter Addition

With this change it would be prudent to add the ability to allow operators to
specify which control plane topology they support. This gives us more guards
against installing layered components on unsupported or unconsidered topologies
like the Master+Arbiter in this enhancement.

This is a nice to have before we go GA and we should be looking into adding it

> to OLM.

Reviewer 14: +1. The OLM team has some thoughts related to this along the lines of generic cluster-provided properties and layered-product provided constraints. This seems to be one use case for that feature. We'll have to think carefully about implementing this in a more general "properties vs. constraints" kind of design.

I don't know that I would make your GA contingent on this feature though since OCP already has multiple topology variants and OLM does not have this feature yet.

One thing we _could_ consider in the meantime is a hint annotation in the CSV that the OCP console could be made aware of. OCP console already supports some of these annotations, and uses them to filter views presented to users.

Author: Awesome, I have it as a nice to have for GA, so we wont let be a critical part of our checklist. Thanks for your input Joe!

This makes the most sense for now since having `InfrastructureTopology` denote a
halfway point between single and three might be a moot distinction for workers.
However, we might need to revisit this as a product team might assume a
`InfrastructureTopology: HighlyAvailable` == `3 Workers`.

### Topology Considerations

#### Hypershift / Hosted Control Planes

We will need to make sure that HyperShift itself can not be installed on the
`HighlyAvailableArbiter` topology. We should also take the opportunity in that
change to also disallow `SingleReplica`.

#### Standalone Clusters

This change is relevant to standalone deployments of OpenShift at the edge or
datacenters. This enhancement specifically deals with this type of deployment.

#### Single-node Deployments or MicroShift

This change does not effect Single Node or MicroShift.

### Implementation Details/Notes/Constraints

Currently there are some behavior unknowns, we will need to test out our POC to
validate some of the desires in this proposal. In it's current version this
proposal is not exhaustive but will be filled out as we implement these goals.

> We currently expect this feature to mainly be used by `baremetal` installs, or

Reviewer 15: Do you plan to prevent it from being used on other platforms? I'm concerned that this breaks many assumptions that CPMS makes about the control plane, and that it will be strictly incompatible. IIRC baremetal doesn't use CPMS at the moment, so perhaps that issue can be avoided for now

specialized hardware that is built to take advantage of this type of
configuration. In the current design we make two paths for cloud and baremetal
installs in the installer. However, the cloud install is primarily for testing,
this might mean that we simplify the installer changes if we are the only ones
using cloud installs, since we can simply alter the manifests in the pipeline
with out needing to change the installer.

> In the current POC, a cluster was created on AWS with 2 masters running

Reviewer 15: Ok, so what was CPMS doing in this cluster? I can't see any discussion of CPMS but we need to think about how we will handle this breaking its assumptions of having three control plane nodes

Reviewer 13: I can answer this, since I've tested this as part of 2NO. It spams the logs about the configuration being invalid, but doesn't take any action.

Reviewer 15: You didn't happen to have anything recorded that captures those logs do you?

Reviewer 13: No, but I could probably get you something like that this afternoon :)

Reviewer 15: If you happen to be testing, and could grab the CPMS logs, I'd like to see exactly what it's complaining about to understand if it's in a safe broken state

Reviewer 13: I think what's going on is that it's never actually being deployed.
The spam happens during bootstrapping, specifically with this error:
```
Dec 10 18:30:06 ip-10-0-100-103 bootkube.sh[4925]: Failed to create "99_openshift-machine-api_master-control-plane-machine-set.yaml" controlplanemachinesets.v1.machine.openshift.io/cluster -n openshift-machine-api: ControlPlaneMachineSet.machine.openshift.io "cluster" is invalid: [spec.replicas: Unsupported value: 2: supported values: "3", "5", <nil>: Invalid value: "null": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation]
```

The control operator logs are happy.
Here's the last 1000 lines of bootstrap + the operator logs.
https://drive.google.com/drive/folders/1-iTWY6GOy6cnukmRzDlkpxES5j3wVZqw

Reviewer 15: Ahh yes, well that makes sense. You'd have to make sure the installer isn't trying to generate a CPMS with the incorrect number of nodes. Which for 2NO means no CPMS. For arbiter, not sure how that will work

Reviewer 13: Egli showed me how to disable CPMS in the manifests, which I think will cover our use case for 2NO.
@eggfoobar will probably have some thoughts on how to handle this properly for arbiter.

Reviewer 15: CPMS also has a generator, that will try to create a CPMS manifest should the cluster not already have one (which serves for upgrades). We probably want a check in there for the topology and have it disable when the topology is not one that it supports 🤔 How is 2no signalled in the ControlPlaneTopology?

Reviewer 13: The plan is to add an enum for DualReplica, following the pattern of SingleReplica.

Reviewer 6: There are no CPMSs for `platform:baremetal` afaik

`m6a.4xlarge` and 1 arbiter running a `m6a.large` instance. At idle the node is
consuming `1.41 GiB / 7.57 GiB	0.135 cores / 2 cores`. The cluster installs and
operates with the following containers running on the arbiter node. This seems
very promising, but we will need to identify any other resources we might want
to run or remove from this list. The list is small already so there might not be
a desire to pre-optimize just yet.

> | Component                                                                                          | Component Outlook on Arbiter                                      |

Reviewer 15: How does this list compare to a regular control plane machine? What's _not_ running on the arbiter, that would be, were it a normal control plane machine?
